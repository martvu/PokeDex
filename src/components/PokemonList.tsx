import { Link } from 'react-router-dom';
import { usePokemonDataList } from '../utils/pokeApi';
import PokemonCard from './PokemonCard';
import { useEffect, useState } from 'react';

export default function PokemonList() {
  const NUM_POKEMON = 151;
  const { data: pokemonDataList, isLoading, error } = usePokemonDataList(NUM_POKEMON);
  const [favoritesArray, setFavoritesArray] = useState<number[]>(
    JSON.parse(localStorage.getItem("favoritesArray") || "[]")
  );
  // handle favoritesArray
  const onToggleFavorite = (pokemonId: number) => {
    if (favoritesArray.includes(pokemonId)) {
      setFavoritesArray(favoritesArray.filter((id:number) => id !== pokemonId));
    } else {
      setFavoritesArray([...favoritesArray, pokemonId]);
    }
  };

  // save favoritesArray
  useEffect(() => {
    localStorage.setItem("favoritesArray", JSON.stringify(favoritesArray));
  }, [favoritesArray]);
  
  if (isLoading) {
    return <div>Loading...</div>;
  }
  if(error) {
    return <div>Error fetching data </div>;
  }
  
  return (
    <div>
      <Link to="/pokemon/1">
        <button>Pokemon Info Page</button>
      </Link>
      <div>
        <div>
          <h1>Pokemon List</h1>
          <div className="pokemon-card-container">
            {pokemonDataList &&
              pokemonDataList.map((pokemon, index) => (
                <PokemonCard key={index + 1} pokemonInfo={pokemon} favoritesArray={favoritesArray} onToggleFavorite={onToggleFavorite}/>
              ))}
          </div>
        </div>
      </div>
    </div>
  );
}

