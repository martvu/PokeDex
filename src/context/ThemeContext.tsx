// ThemeContext.tsx
import { createContext, useState, ReactNode } from 'react';

interface ThemeContextType {
  darkMode: boolean;
  toggleDarkMode: () => void;
};

export const ThemeContext = createContext<ThemeContextType | undefined>(
  undefined
);

export function ThemeProvider({ children }: { children: ReactNode }) {
  const [darkMode, setDarkMode] = useState(
    localStorage.getItem('darkMode') === 'true'
  );

  const toggleDarkMode = () => {
    setDarkMode(prevDarkMode => !prevDarkMode);
    localStorage.setItem('darkMode', JSON.stringify(!darkMode));
  };

  return (
    <ThemeContext.Provider value={{ darkMode, toggleDarkMode }}>
      {children}
    </ThemeContext.Provider>
  );
}
