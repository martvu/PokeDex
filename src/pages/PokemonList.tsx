import { usePokemonDataList } from '../utils/pokeApi';
import PokemonCard from '../components/PokemonCard';
import { createContext, useState } from 'react';
import { Pokemon } from '../utils/pokeApiTypes';
import Header from '../components/Header';
import Filters from '../components/Filters';
import loadingIcon from '../assets/img/poke-loading.gif';

export interface FavoriteContextProps {
  favoritesArray: number[];
  setFavoritesArray: (favoritesArray: number[]) => void;
}

export const FavoriteContext = createContext<FavoriteContextProps | undefined>(
  undefined
);

export default function PokemonList() {
  const NUM_POKEMON = 151;
  const {
    data: pokemonDataList,
    isLoading,
    error,
  } = usePokemonDataList(NUM_POKEMON);
  // handle filters
  const [displayedPokemonList, setDisplayedPokemonList] = useState<Pokemon[]>(
    []
  );
  const [favoritesArray, setFavoritesArray] = useState<number[]>(
    JSON.parse(localStorage.getItem('favoritesArray') || '[]')
  );

  if (error) {
    return <div>Error fetching data </div>;
  }

  return (
    <FavoriteContext.Provider value={{ favoritesArray, setFavoritesArray }}>
      <Header />
      <Filters
        favoritesArray={favoritesArray}
        pokemonDataList={pokemonDataList as Pokemon[]}
        setDisplayedPokemonList={setDisplayedPokemonList}
      />
      {isLoading && (
        <div className={`loading-screen pokemon-text`}>
          <img src={loadingIcon} alt="loading icon" />
        </div>
      )}
      {(error as Error) && (
        <div className="error-screen pokemon-text">
          <h3>An error has occured</h3>
        </div>
      )}
      {!error && !isLoading && displayedPokemonList?.length === 0 && (
        <div className="no-results pokemon-text">
          <h3>No Pokemon Found!</h3>
        </div>
      )}
      {!isLoading && (
        <ul className="pokemon-list-container">
          {displayedPokemonList?.map((pokemon: Pokemon) => (
            <PokemonCard key={pokemon.id} pokemonDetails={pokemon} />
          ))}
        </ul>
      )}
    </FavoriteContext.Provider>
  );
}
