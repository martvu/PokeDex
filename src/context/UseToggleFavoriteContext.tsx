import { Pokemon } from '../utils/pokeApiTypes.tsx';
import { useContext, useEffect } from 'react';
import { FavoriteContext } from '../pages/PokemonList.tsx';

export function useToggleFavoriteContext(pokemonDetails: Pokemon) {
  const context = useContext(FavoriteContext);

  if (!context) {
    throw new Error(
      'useToggleFavoriteContext must be used within a FavoriteContextProvider'
    );
  }
  const { favoritesArray, setFavoritesArray } = context;

  // handle favoritesArray
  const onToggleFavorite = (pokemonId: number) => {
    if (favoritesArray.includes(pokemonId)) {
      setFavoritesArray(
        favoritesArray.filter((id: number) => id !== pokemonId)
      );
    } else {
      setFavoritesArray([...favoritesArray, pokemonId]);
    }
  };

  // save favoritesArray
  useEffect(() => {
    localStorage.setItem('favoritesArray', JSON.stringify(favoritesArray));
  }, [favoritesArray]);

  return {
    toggleFavorite: (pokemonId: number) => {
      onToggleFavorite(pokemonId); // Call the onToggleFavorite function from the context
    },
    isFavorite: favoritesArray.includes(pokemonDetails.id), // Check if the current Pokemon is a favorite
    allFavorites: favoritesArray, // Return the favoritesArray
  };
}
