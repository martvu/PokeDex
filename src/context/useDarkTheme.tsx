import { useContext } from 'react';
import { ThemeContext } from './ThemeContext';

function useDarkTheme() {
  const context = useContext(ThemeContext);
  if (!context) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
}

export default useDarkTheme;
