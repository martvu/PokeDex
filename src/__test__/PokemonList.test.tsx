import { render, screen } from '@testing-library/react';
import { describe, expect, test } from 'vitest';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import PokemonList from '../components/PokemonList';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import { ThemeProvider } from '../context/ThemeContext';

const queryClient = new QueryClient();

describe('PokemonList', () => {
  test('pokemonlist renders', async () => {
    render(
      <QueryClientProvider client={queryClient}>
        <ThemeProvider>
          <MemoryRouter initialEntries={['/project1']}>
            <Routes>
              <Route path='/project1' element={<PokemonList />} />
            </Routes>
          </MemoryRouter>
        </ThemeProvider>
      </QueryClientProvider>
    );

    expect(await screen.findByText(/bulbasaur/i)).toBeInTheDocument();
    expect(await screen.findByText(/ivysaur/i)).toBeInTheDocument();
    expect(await screen.findByText(/#001/i)).toBeInTheDocument();
    expect(await screen.findByText(/#002/i)).toBeInTheDocument();
    expect(await screen.findByText(/type/i)).toBeInTheDocument();
    expect(await screen.findByText(/pokedex/i)).toBeInTheDocument();
  });
});

describe('Snapshot PokemonList', () => {
  test('snapshot of pokemonlist', async () => {
    const { asFragment } = render(
      <QueryClientProvider client={queryClient}>
        <ThemeProvider>
          <MemoryRouter initialEntries={['/project1']}>
            <Routes>
              <Route path="/project1" element={<PokemonList />} />
            </Routes>
          </MemoryRouter>
        </ThemeProvider>
      </QueryClientProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

