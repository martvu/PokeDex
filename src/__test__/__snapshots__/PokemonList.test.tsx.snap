// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot PokemonList > snapshot of pokemonlist 1`] = `
<DocumentFragment>
  <header
    class="header"
  >
    <div
      class="pokeball-header-icon"
    >
      <img
        alt="pokeball icon"
        src="/project1/src/assets/img/pb-icon.svg"
      />
      <div
        class="header-link-container"
      >
        <a
          class="link"
          href="/project1"
        >
          <h1>
            Pokedex
          </h1>
        </a>
      </div>
      <span
        class="switch-container"
      >
        <button
          class="theme-switch "
        >
          Turn Dark
        </button>
      </span>
    </div>
  </header>
  <div
    class="filter-container"
  >
    <div
      class="search-bar-container"
    >
      <svg
        class="search-icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 18a7.952 7.952 0 0 0 4.897-1.688l4.396 4.396 1.414-1.414-4.396-4.396A7.952 7.952 0 0 0 18 10c0-4.411-3.589-8-8-8s-8 3.589-8 8 3.589 8 8 8zm0-14c3.309 0 6 2.691 6 6s-2.691 6-6 6-6-2.691-6-6 2.691-6 6-6z"
        />
        <path
          d="M11.412 8.586c.379.38.588.882.588 1.414h2a3.977 3.977 0 0 0-1.174-2.828c-1.514-1.512-4.139-1.512-5.652 0l1.412 1.416c.76-.758 2.07-.756 2.826-.002z"
        />
      </svg>
      <input
        class="search-bar"
        placeholder="Search Pokemon"
        type="text"
      />
    </div>
    <button
      class="filter-favorite-btn "
      title="Show Favorites"
    >
      <svg
        class="favorite-icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
        />
      </svg>
    </button>
    <button
      class="filter-reset-btn"
      title="Reset Filters"
    >
      <svg
        class="reset-icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20,8 C18.5974037,5.04031171 15.536972,3 12,3 C7.02943725,3 3,7.02943725 3,12 C3,16.9705627 7.02943725,21 12,21 L12,21 C16.9705627,21 21,16.9705627 21,12 M21,3 L21,9 L15,9"
          fill="none"
          stroke="#000"
          stroke-width="2"
        />
      </svg>
    </button>
    <div
      class="filter-options-container"
    >
      <div
        class="filter-select"
      >
        <label
          class="filter-text"
          for="type-select"
        >
          Type:
        </label>
        <select
          id="type-select"
        >
          <option
            value="all"
          >
            All
          </option>
          <option
            value="normal"
          >
            Normal
          </option>
          <option
            value="fire"
          >
            Fire
          </option>
          <option
            value="water"
          >
            Water
          </option>
          <option
            value="electric"
          >
            Electric
          </option>
          <option
            value="grass"
          >
            Grass
          </option>
          <option
            value="ice"
          >
            Ice
          </option>
          <option
            value="fighting"
          >
            Fighting
          </option>
          <option
            value="poison"
          >
            Poison
          </option>
          <option
            value="ground"
          >
            Ground
          </option>
          <option
            value="flying"
          >
            Flying
          </option>
          <option
            value="psychic"
          >
            Psychic
          </option>
          <option
            value="bug"
          >
            Bug
          </option>
          <option
            value="rock"
          >
            Rock
          </option>
          <option
            value="ghost"
          >
            Ghost
          </option>
          <option
            value="dragon"
          >
            Dragon
          </option>
          <option
            value="dark"
          >
            Dark
          </option>
          <option
            value="steel"
          >
            Steel
          </option>
          <option
            value="fairy"
          >
            Fairy
          </option>
        </select>
      </div>
      <div
        class="filter-select"
      >
        <label
          class="filter-text"
          for="sort-select"
        >
          Sort By
        </label>
        <select
          id="sort-select"
        >
          <option
            value="ID"
          >
            ID
          </option>
          <option
            value="A-Z"
          >
            A-Z
          </option>
          <option
            value="high to low"
          >
            High to low
          </option>
        </select>
      </div>
    </div>
  </div>
  <div
    class="pokemon-list"
  >
    <ul
      class="pokemon-list-container"
    >
      <div
        class="pokemon-card-container"
      >
        <div
          class="poketext pokemon-name"
        >
          Bulbasaur
        </div>
        <div
          class="poketext pokemon-id"
        >
          #001
        </div>
        <div
          class="favorite-button"
        >
          <svg
            class="favorite-icon-inactive"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
            />
          </svg>
        </div>
        <div
          class="sprite-bg-container"
        >
          <img
            alt="pokeball background"
            class="sprite-bg"
            loading="lazy"
            src="/project1/src/assets/img/pb-icon.svg"
          />
        </div>
        <a
          href="/project1/pokemon/1"
        >
          <div
            class="sprite-container"
            data-testid="link-infopage"
          >
            <img
              alt="bulbasaur"
              class="pokemon-sprite"
              loading="lazy"
              src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/1.png"
            />
          </div>
        </a>
        <div
          class="type-container"
        >
          <span
            class="poketext type-icon"
            style="background-color: rgb(122, 199, 76);"
          >
            Grass
          </span>
          <span
            class="poketext type-icon"
            style="background-color: rgb(163, 62, 161);"
          >
            Poison
          </span>
        </div>
      </div>
      <div
        class="pokemon-card-container"
      >
        <div
          class="poketext pokemon-name"
        >
          Ivysaur
        </div>
        <div
          class="poketext pokemon-id"
        >
          #002
        </div>
        <div
          class="favorite-button"
        >
          <svg
            class="favorite-icon-inactive"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
            />
          </svg>
        </div>
        <div
          class="sprite-bg-container"
        >
          <img
            alt="pokeball background"
            class="sprite-bg"
            loading="lazy"
            src="/project1/src/assets/img/pb-icon.svg"
          />
        </div>
        <a
          href="/project1/pokemon/2"
        >
          <div
            class="sprite-container"
            data-testid="link-infopage"
          >
            <img
              alt="ivysaur"
              class="pokemon-sprite"
              loading="lazy"
              src="#"
            />
          </div>
        </a>
        <div
          class="type-container"
        >
          <span
            class="poketext type-icon"
            style="background-color: rgb(122, 199, 76);"
          >
            Grass
          </span>
          <span
            class="poketext type-icon"
            style="background-color: rgb(163, 62, 161);"
          >
            Poison
          </span>
        </div>
      </div>
    </ul>
  </div>
</DocumentFragment>
`;
